prov/util/src/util_mem_hooks.c

```diff
@@ -125,246 +154,209 @@ static struct ofi_intercept intercepts[] = {
-#define OFI_ELF_R_SYM ELF64_R_SYM
-#else
-#define OFI_ELF_R_SYM ELF32_R_SYM
 #endif
 
-static void *ofi_dl_func_addr(ElfW(Addr) base, const ElfW(Phdr) *phdr,
-			      int16_t phnum, int phent, const char *symbol)
-{
-	const ElfW(Phdr) *dphdr;
-	ElfW(Rela) *reloc;
-	void *jmprel, *strtab;
-	char *elf_sym;
-	uint32_t relsymidx;
-	ElfW(Sym) *symtab;
-	size_t pltrelsz;
-
-	dphdr = ofi_get_phdr_dynamic(phdr, phnum, phent);
-	jmprel = ofi_get_dynentry(base, dphdr, DT_JMPREL);
-	symtab = (ElfW(Sym) *) ofi_get_dynentry(base, dphdr, DT_SYMTAB);
-	strtab = ofi_get_dynentry (base, dphdr, DT_STRTAB);
-	pltrelsz = (uintptr_t) ofi_get_dynentry(base, dphdr, DT_PLTRELSZ);
-
-	for (reloc = jmprel; (intptr_t) reloc < (intptr_t) jmprel + pltrelsz;
-	     reloc++) {
-		relsymidx = OFI_ELF_R_SYM(reloc->r_info);
-		elf_sym = (char *) strtab + symtab[relsymidx].st_name;
-		if (!strcmp(symbol, elf_sym))
-			return (void *) (base + reloc->r_offset);
-        }
-
-        return NULL;
-}
-
-static int ofi_intercept_dl_calls(ElfW(Addr) base, const ElfW(Phdr) *phdr,
-				  const char *phname, int16_t phnum, int phent,
-				  struct ofi_intercept *intercept)
+static inline int ofi_write_patch(unsigned char *patch_data, void *address,
+				  size_t data_size)
 {
-	struct ofi_dl_intercept *dl_entry;
 	long page_size = ofi_get_page_size();
-	void **func_addr, *page;
-	int ret;
+	void *base = ofi_get_page_start(address, page_size);
+	void *bound = ofi_get_page_end(address, page_size);
+	size_t length = (uintptr_t) bound - (uintptr_t) base;
 
-	FI_DBG(&core_prov, FI_LOG_MR,
-	       "intercepting symbol %s from dl\n", intercept->symbol);
-	func_addr = ofi_dl_func_addr(base, phdr, phnum, phent, intercept->symbol);
-	if (!func_addr)
-		return FI_SUCCESS;
-
-	page = (void *) ((intptr_t) func_addr & ~(page_size - 1));
-	ret = mprotect(page, page_size, PROT_READ | PROT_WRITE);
-	if (ret < 0)
-		return -FI_ENOSYS;
-
-	if (*func_addr != intercept->our_func) {
-		dl_entry = malloc(sizeof(*dl_entry));
-		if (!dl_entry)
-			return -FI_ENOMEM;
-
-		dl_entry->dl_func_addr = func_addr;
-		dl_entry->dl_func = *func_addr;
-		*func_addr = intercept->our_func;
-		dlist_insert_tail(&dl_entry->entry, &intercept->dl_intercept_list);
+	if (mprotect(base, length, PROT_EXEC|PROT_READ|PROT_WRITE)) {
+		FI_WARN(&core_prov, FI_LOG_MR,
+			"mprotect to set PROT_WRITE on %p len %lu failed: %s\n",
+			(void *) base, length, strerror(errno));
+		return -errno;
 	}
 
-	return FI_SUCCESS;
-}
+	memcpy(address, patch_data, data_size);
 
-static int ofi_intercept_phdr_handler(struct dl_phdr_info *info,
-                                    size_t size, void *data)
-{
-	struct ofi_intercept *intercept = data;
-	int phent, ret;
+	ofi_clear_instruction_cache((uintptr_t) address, data_size);
 
-	phent = getauxval(AT_PHENT);
-	if (phent <= 0) {
-		FI_DBG(&core_prov, FI_LOG_MR, "failed to read phent size");
-		return -FI_EINVAL;
-	}
+	/*
+	 * Nothing we can do here if this fails so ignore the return code. It
```

Comment:
I don't follow the logic.  If it fails, isn't there a problem?

Reply:
There is a problem - the page will still be writable. I'm not sure there's anything we can do here besides print a message.